### PR TITLE
Ristretto Implementation - A high performance memory-bound Go cache 

### PR DESCRIPTION
**Pull request by rbo13**
Hello Team,

This PR adds an implementation to [ristretto](https://github.com/dgraph-io/ristretto) - a high-performance memory-bound Go cache.

### CODE REVIEW

**Comment by ReneWerner87** [comment]
thx for the contribution

please add the storage in the dependabot.yml

**Comment by rbo13** [comment]
Hello @ReneWerner87. I have added the following changes requested. Please check my new PR 😄. Thanks 👍🏽

File: ristretto/ristretto.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+func New(config ...Config) *Storage {
+	cfg := configDefault(config...)
+	cache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters: cfg.NumCounters,
+		MaxCost:     cfg.MaxCost,
+		BufferItems: cfg.BufferItems,
+	})
+
+	if err != nil {
+		return nil
```
Comment (hi019) [inline ristretto/ristretto.go:25]: We should send this error to the program somehow. With the other storages we panic, but returning `*Storage, error` and letting the programmer handle the error is probably better. But we should also be consistent :)

@ReneWerner87 what do you think?

File: ristretto/ristretto.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	buf, asserted := item.([]byte)
+	if !asserted {
+		return nil, nil
+	}
+
+	return buf, nil
+}
+
+// Set stores the given value for the given key along
+// with an expiration value, 0 means no expiration.
```
Comment (hi019) [inline ristretto/ristretto.go:57]:
```suggestion
// with an expiration value, time.Time{} means no expiration.
```

File: ristretto/ristretto.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+func (s *Storage) Set(key string, val []byte, exp time.Duration) error {
+	if len(key) <= 0 || len(val) <= 0 {
+		return nil
+	}
+	saved := s.cache.SetWithTTL(key, val, s.defaultCost, exp)
+	if !saved {
+		return nil
+	}
+	// wait for value to pass through buffers
+	time.Sleep(10 * time.Millisecond)
```
Comment (hi019) [inline ristretto/ristretto.go:68]: Can you explain a bit more what this is for?

File: ristretto/ristretto_test.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	var (
+		key = "john"
+		val = []byte("doe")
+		exp = 1 * time.Second
+	)
+
+	err := testStore.Set(key, val, exp)
+	utils.AssertEqual(t, nil, err)
+
+	time.Sleep(1100 * time.Millisecond)
```
Comment (hi019) [inline ristretto/ristretto_test.go:60]: I assume this line suspends the rest of the tests until the value is expired? What about using `testStore.Reset()` instead?

**Review by hi019 — CHANGES_REQUESTED** [review]
Thank you for the PR!

File: ristretto/ristretto.go (hunk 3 above)
Comment (rbo13) [reply ristretto/ristretto.go:68]: Actually, I don't really know what is this for. It is stated in the documentation of ristretto here is the example [link](https://github.com/dgraph-io/ristretto#example). But I believe it is safe to delete this line as it is not important

File: ristretto/ristretto_test.go (hunk 4 above)
Comment (rbo13) [reply ristretto/ristretto_test.go:60]: Yeah your assumption is correct, we can just simply remove this one as `testStore.Reset()` will do the same.

File: ristretto/ristretto.go (hunk 1 above)
Comment (ReneWerner87) [reply ristretto/ristretto.go]: Exactly, good found

Comment (rbo13) [reply ristretto/ristretto.go]: I have added the panic line to provide consistency.

**Comment by hi019** [comment]
Going to pull your repo and test it out later today. PS please mark review conversations as resolved when appropriate

**Comment by rbo13** [comment]
> Going to pull your repo and test it out later today. PS please mark review conversations as resolved when appropriate

Thanks for merging my work @hi019 😄 my bad for not marking the conversation as resolved. Will surely do it next time I contribute to this project 😃 again, thank you so much for approving my contribution. I really appreciate it.

**Pull request merged**